crates/jit/src/object.rs

```diff
@@ -0,0 +1,338 @@
+    );
+
+    let mut unwind_info = Vec::new();
+
+    // Create symbols for imports -- needed during linking.
+    let mut func_symbols = PrimaryMap::with_capacity(compilation.len());
+    for index in 0..module.local.num_imported_funcs {
+        let symbol_id = obj.add_symbol(Symbol {
+            name: func_symbol_name(FuncIndex::new(index)).as_bytes().to_vec(),
+            value: 0,
+            size: 0,
+            kind: SymbolKind::Text,
+            scope: SymbolScope::Linkage,
+            weak: false,
+            section: SymbolSection::Undefined,
+            flags: SymbolFlags::None,
+        });
+        func_symbols.push(symbol_id);
+    }
+    // Create symbols and section data for the compiled functions.
+    for (index, func) in compilation.into_iter().enumerate() {
+        let off = obj.append_section_data(section_id, &func.body, 1);
+        let symbol_id = obj.add_symbol(Symbol {
+            name: func_symbol_name(module.local.func_index(DefinedFuncIndex::new(index)))
+                .as_bytes()
+                .to_vec(),
+            value: off,
+            size: func.body.len() as u64,
+            kind: SymbolKind::Text,
+            scope: SymbolScope::Compilation,
+            weak: false,
+            section: SymbolSection::Section(section_id),
+            flags: SymbolFlags::None,
+        });
+        func_symbols.push(symbol_id);
+        if let Some(UnwindInfo::WindowsX64(info)) = &func.unwind_info {
+            // Windows prefers Unwind info after the code -- writing it here.
+            let unwind_size = info.emit_size();
+            let mut unwind_info = vec![0; unwind_size];
+            info.emit(&mut unwind_info);
+            let _off = obj.append_section_data(section_id, &unwind_info, 4);
+        }
+        // Preserve function unwind info.
+        if let Some(info) = &func.unwind_info {
+            unwind_info.push(ObjectUnwindInfo::Func(
+                FuncIndex::new(module.local.num_imported_funcs + index),
+                info.clone(),
+            ))
+        }
+    }
+
+    let mut trampoline_relocs = HashMap::new();
+    let mut cx = FunctionBuilderContext::new();
+    // Build trampolines for every signature.
+    for (i, (_, native_sig)) in module.local.signatures.iter() {
+        let (func, relocs) =
+            build_trampoline(isa, &mut cx, native_sig, std::mem::size_of::<u128>())?;
+        let off = obj.append_section_data(section_id, &func.body, 1);
+        let symbol_id = obj.add_symbol(Symbol {
+            name: trampoline_symbol_name(i).as_bytes().to_vec(),
+            value: off,
+            size: func.body.len() as u64,
+            kind: SymbolKind::Text,
+            scope: SymbolScope::Compilation,
+            weak: false,
+            section: SymbolSection::Section(section_id),
+            flags: SymbolFlags::None,
+        });
+        trampoline_relocs.insert(symbol_id, relocs);
+        if let Some(UnwindInfo::WindowsX64(info)) = &func.unwind_info {
```

Comment:
This has a similar branch above for normal code symbols, so could these two perhaps be unified with a helper function?